[Ng Seng Leng] iP

ngsengleng opened the pull request:
### DukePro

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's

* text-based
* easy to learn
* ~FAST~ *SUPER* FAST to use

All you need to do is, 

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And its **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Add Gradle support

ngsengleng committed:
add features for level 1

ngsengleng committed:
add features for level 2

ngsengleng committed:
add features for level 3 using a Task class

ngsengleng committed:
add features for level 4

ngsengleng committed:
change accessibility of some variables in Event and Deadline class

ngsengleng committed:
add testing for existing features

ngsengleng committed:
change some string formatting to String.format

ngsengleng committed:
add error handling and modify some string formatting for certain features

ngsengleng committed:
add features for level 6

ngsengleng committed:
Add save feature for duke and classes to handle file read/write

ngsengleng committed:
Add save feature for level 7 and update tests

ngsengleng committed:
change duke to recognise proper date format

ngsengleng committed:
Merge branch 'branch-Level-8'

Merge level 8 features into main branch

ngsengleng committed:
merge branches 7 and 8 and change string output

ngsengleng committed:
abstract some functions out to individual classes (OOP)

ngsengleng committed:
refactor classes into one package (duke)

ngsengleng committed:
Add junit tests for some classes in duke package

ngsengleng committed:
include meta file for jar binary

ngsengleng committed:
Clean up javadoc comments

ngsengleng committed:
Change names of some List variables to match assigned Java convention

ngsengleng committed:
Add find function to the bot

ngsengleng committed:
Merge branch 'branch-Level-9'

ngsengleng committed:
Merge branch 'branch-A-CodingStandard'

ngsengleng committed:
add header comments to classes

ngsengleng committed:
Merge branch 'add-gradle-support'

Leofeng10 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,231 @@
+        String command = parser.getCommand(input);
+        try {
+            switch(command) {
+                case "items":
+                    ui.outputWrapper(listOfItems);
+                    break;
+                case "completed":
+                    markAsDone(input);
+                    break;
+                case "todo":
+                    addAsTodo(input);
+                    break;
+                case "event":
+                    addAsEvent(input);
+                    break;
+                case "deadline":
+                    addAsDeadline(input);
+                    break;
+                case "delete":
+                    deleteItem(input);
+                    break;
+                case "find":
+                    findItems(input);
+                    break;
+                default:
+                    markAsInvalid(input);
+                    break;
```
should the indentation reduce a tap space?

ngsengleng replied:
yep thanks i missed this part out

Leofeng10 commented on src/main/java/duke/FileParser.java:
```diff
@@ -0,0 +1,53 @@
+                    task = new Todo(taskDescription);
+                    break;
+                default:
+                    throw new DukeException("a line could not be read from the savefile or was read incorrectly. " +
+                            "An empty list will be used instead.");
```
should we put the "+" at the front of next line?

ngsengleng replied:
thanks for pointing it out

Leofeng10 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+    public Event(String description, String date) {
+        super(description);
+        try {
+            int[] dateArr = Arrays.stream(date.split("-")).mapToInt(Integer::parseInt).toArray();
+            this.date = LocalDate.of(dateArr[0], dateArr[1], dateArr[2]);
```
is it better to break these two line into smaller parts or add some comments to increase readability?

ngsengleng replied:
yep, i added some comments for clarity in an upcoming commit

Leofeng10 reviewed (commented):
Great work done. Nice code structure. Very readable code! 👍
Your code follows the guideline and coding standard really well. The problems I found were just some small issues. Nice work!

ngsengleng committed:
add support for testing and building using gradle

ngsengleng committed:
add support for testing and building using gradle

crypto-code commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+
+    private LocalDate date;
+
+    public Deadline(String description, String date) {
```
This may feel a bit cluttered, you could consider making a separate method to process the Date

Jai2501 replied:
I agree, having the constructor call a method like `parseDate()` would be better.

crypto-code commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+    protected LocalDate date;
+
+    public Event(String description, String date) {
+        super(description);
```
Even this could be implemented in a separate method.

crypto-code commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,231 @@
+    private void handleInput(String input) {
+        String command = parser.getCommand(input);
+        try {
+            switch(command) {
```
According to Java Coding standards, for a switch statement, case clauses aren't to be intended. I blame IntelliJ for this. You could try to fix this manually.

ngsengleng replied:
yep i fixed the indentations

crypto-code reviewed (commented):
Good Job with the Code 👍

ngsengleng committed:
add support for javafx gui

ngsengleng committed:
Modify javadocs to specifications of new functions

ngsengleng committed:
Fix merge conflicts

ngsengleng committed:
Fix merge conflicts

ngsengleng committed:
Attach save function to bye command

adam-ky commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,218 @@
+    private String markAsDone(String input){
+        try {
+            String[] keywords = input.split(" ");
+            String command = keywords[0];
```
Is this command variable used?

ngsengleng replied:
thanks i missed out this line

adam-ky commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+        String timestamp = String.format("(at: %s)", date.toString());
+
+        if (isDone) {
+            return eventMarker + "|" + hasCross + "|" + item + "|" + date.toString();
```
toString() here might not be necessary?

adam-ky commented on src/main/java/duke/FileParser.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            String[] args = input.split("\\|");
+            String taskType = args[0];
+            Boolean isCompleted = args[1].contains("X");
```
Maybe a primitive boolean can be used instead of the wrapper version?

adam-ky commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+
```
Minor formatting issue of an empty newline 😄

adam-ky commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,218 @@
+                    return findItems(input);
+                case "bye":
+                    try {
+                        storage.save(listOfItems);
```
Perhaps this try-catch block of IOException could go under Storage's save method instead?

ngsengleng replied:
yep that sounds like a better idea!

adam-ky commented on src/main/java/duke/FileParser.java:
```diff
@@ -0,0 +1,53 @@
+     * @return A Task object to be added to the task list.
+     * @throws DukeException Throws an exception when an error is encountered when parsing the line from file.
+     */
+    public Task parse(String input) throws DukeException {
```
Code is clean and easy to understand, good job! 🥳

adam-ky commented on src/main/java/duke/FileParser.java:
```diff
@@ -0,0 +1,53 @@
+            String taskDescription = args[2];
+            String datetime = args.length > 3 ? args[3] : "";
+            Task task;
+            switch(taskType) {
```
Note that the case should be indented to the same line as the switch statement.

adam-ky commented on src/test/java/duke/FileParserTest.java:
```diff
@@ -0,0 +1,33 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class FileParserTest {
```
Tests are well written and follow correct coding standards! 💯

adam-ky reviewed (commented):
Overall good job and keep it up! Your code follows the coding standards for the most part 😃

Jai2501 commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,60 @@
+        db.flip();
+        return db;
+    }
+}
```
I think you should have a newline at the end?

Jai2501 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,218 @@
+            String keyword = parser.getDescription(input);
+            Tasklist validItems = listOfItems.findAllBy(keyword);
+            return ui.listToPrintableString(validItems);
+        }catch (ArrayIndexOutOfBoundsException e){
```
You may add a space between the `}` and the `catch` statement as well as `(` and `{`.
```suggestion
        } catch (ArrayIndexOutOfBoundsException e) {
```

Jai2501 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,218 @@
+    public String getResponse(String input) {
+        String command = parser.getCommand(input);
+        try {
+            switch(command) {
```
I think the `case` clauses for a `switch` statement aren’t supposed to be indented.

Jai2501 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+        super(description);
+        try {
+            int[] dateArr = Arrays.stream(date.split("-")).mapToInt(Integer::parseInt).toArray();
+            this.date = LocalDate.of(dateArr[0], dateArr[1], dateArr[2]);
```
I think it would be better if we made a separate function for parsing and obtaining the date as mentioned previously.

Jai2501 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,57 @@
+package duke;
```
Well written code with intuitive variable names.

Jai2501 commented on src/main/resources/view/DialogBox.fxml:
```diff
@@ -0,0 +1,16 @@
+    <padding>
+        <Insets bottom="15.0" left="5.0" right="5.0" top="15.0" />
+    </padding>
+</fx:root>
```
Some of the files had this minor issue, where they did not have a newline at the end of the file.

Jai2501 reviewed (commented):
All in all, the code is clean and well written. There are no readability issues. However, some instances were found where there were issues with the code quality and the coding standard. These few nits need to be fixed. LGTM.

ngsengleng committed:
Refactor code to fit java coding standards

Constructors for Deadline and Event are too cluttered.

Extracting lines of code out into methods results in better code clarity and better
readability.

Let's extract the lines for parsing date in string format into LocalDate format into a
private method makes the constructor look much cleaner and less cluttered.

Putting the date parsing into a method makes the code more readable and cleaner.

ngsengleng committed:
Refactor code to fit java coding standards

Constructors for Deadline and Event are too cluttered.

Extracting lines of code out into methods results in better code clarity and better
readability.

Extracting the lines for parsing date in string format into LocalDate format into a
private method makes the constructor look much cleaner and less cluttered.

Putting the date parsing into a method makes the code more readable and cleaner.

ngsengleng committed:
Fix minor coding standard infractions

ngsengleng committed:
Fix minor coding standard infractions

ngsengleng committed:
Add assertions for Deadline and Event class methods

Current methods have no assertions to ensure proper flow of
execution.

Adding assertions ensure that no variables are null when passed
into the method to ensure no unexpected outputs.

Let's add assertions to check for null variables.

This checks for erroneous function inputs and ensures that the
code is executed the way it is intended to.

ngsengleng committed:
Shift parseDate method in Deadline and Event class to Task parent class

Both Deadline and Event class have the same method for parsing a date in
string format.

This breaks the DRY principle as there is a duplication of methods in the
code base.

Let's shift the method to their parent class Task so the method is inherited
instead.

Inheriting the method ensures that there is no duplication of code and the
DRY principle is adhered to.

ngsengleng committed:
Merge branch 'master' into branch-A-Assertions

ngsengleng committed:
Merge pull request #1 from ngsengleng/branch-A-Assertions

Branch a assertions

ngsengleng committed:
Fix merge conflicts with master

ngsengleng committed:
Merge pull request #2 from ngsengleng/branch-A-CodeQuality

Branch a code quality

ngsengleng committed:
Fix code to make it more readable

ngsengleng committed:
Add github actions workflow file

ngsengleng committed:
Change code to fit checkstyle recommendations

ngsengleng committed:
Change junit tests to pass checkstyle

ngsengleng committed:
Add command to request for list of commands in Duke

ngsengleng committed:
Change listOfCommands to private access

ngsengleng committed:
Change listOfCommands to final

ngsengleng committed:
Fix bug where list of items would not be fully shown if too long

ngsengleng committed:
Modify README

ngsengleng committed:
Add images to user guide

ngsengleng committed:
Set theme jekyll-theme-slate

ngsengleng committed:
Update README.md

ngsengleng committed:
Change display message for empty task list

ngsengleng committed:
Merge branch 'master' of https://github.com/ngsengleng/ip

ngsengleng committed:
Add help feature to user guide

ngsengleng committed:
Update README.md

ngsengleng committed:
Update README.md
